static/js/controllers/configuration-forms.js

```diff
@@ -28,24 +30,102 @@ var moment = require('moment'),
 
       'ngInject';
 
-      $scope.uploading = false;
+      $scope.json = { uploading: false };
+      $scope.xform = { uploading: false };
 
-      var uploadFinished = function(err) {
-        $scope.uploading = false;
-        $scope.error = !!err;
-        $scope.success = !err;
+      var uploadJsonFinished = function(err) {
+        uploadFinished('json', err);
+      };
+
+      var uploadXFormFinished = function(err) {
+        uploadFinished('xform', err);
+      };
+
+      var uploadFinished = function(type, err) {
+        $scope[type] = {
+          uploading: false,
+          error: !!err,
+          success: !err,
+        };
         if (err) {
           $log.error('Upload failed', err);
         }
       };
 
-      var uploadForms = function() {
-        $scope.uploading = true;
-        $scope.error = false;
-        $scope.success = false;
-        var files = $('#forms-upload-form .uploader')[0].files;
+      var uploadXForm = function() {
+        $scope.xform = {
+          uploading: true,
+          error: false,
+          success: false,
+        };
+
+        var formFiles = $('#forms-upload-xform .form.uploader')[0].files;
+        if (!formFiles || formFiles.length === 0) {
+          uploadXFormFinished(new Error('XML file not found'));
+        }
+
+        var contextFiles = $('#forms-upload-xform .context.uploader')[0].files;
+        if (!contextFiles || contextFiles.length === 0) {
+          uploadXFormFinished(new Error('JSON context file not found'));
+        }
+
+        $q.all([
+            FileReader(formFiles[0]),
+            FileReader(contextFiles[0]).then(JsonParse) ])
+          .then(function(results) {
+            var xml = results[0];
+            var context = results[1];
+
+            var $xml = $($.parseXML(xml));
+            var title = $xml.find('title').text();
+            var formId = $xml.find('instance').children().first().attr('id');
+
+            var update = function(doc) {
+              doc.context = context;
+              doc.title = title;
+              doc.type = 'form';
+              doc.internalId = formId;
+              if (!doc._attachments) {
+                doc._attachments = {};
+              }
+              doc._attachments.xml = {
+                content_type: 'application/xml',
+                data: new Blob([xml]),
+              };
+              return doc;
+            };
+
+            var couchId = 'form:' + formId;
+            DB().get(couchId, { include_attachments:true })
```

Comment:
This block could be cleaned up a lot. Try this...

```js
DB().get(couchId, { include_attachments:true })
  .catch(function(err) {
    if (err.status === 404) {
      return { _id: couchId };
    }
    throw err;
  })
  .then(function(doc) {
    return DB().put(update(doc));
  })
  .then(function() {
    uploadXFormFinished();
  })
  .catch(uploadXFormFinished);
```

I think this is functionally equivalent and I find it much easier to read (less nesting, fewer exit points).